float32 output_val